File: src/rac/services/improve.py
"""Advisory, schema-driven completion guidance (`rac improve`).

`rac improve <file>` reports which required and recommended sections an artifact
is missing and, for each, the schema-defined questions that guide filling it in.
It is strictly read-only (REQ-004) and deterministic (ADR-002): no AI, no
rewriting, no content generated beyond schema-derived placeholders.

Guidance depends only on the artifact *type* and a *schema comparison*
(:func:`rac.core.classification.missing_sections`) — never on classification
internals, and it never feeds back into classification, validation, or
statistics. A type earns support purely through its :class:`ArtifactSpec`: it
must define guidance for every expected section (:func:`supports_improve`).
There is no per-type engine — the same pipeline serves every artifact family,
which is why a new type becomes improvable the moment its spec gains complete
guidance.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from rac.core.artifacts import ArtifactSpec, spec_for
from rac.core.classification import classify, missing_sections
from rac.core.markdown import parse, parse_file
from rac.core.models import Product


def supports_improve(spec: ArtifactSpec) -> bool:
    """True when ``spec`` defines guidance for every section it expects.

    This gate is what makes support spec-driven rather than hard-coded: a future
    type cannot become improvable until its schema covers all of its required
    and recommended sections.
    """
    return all(section in spec.guidance for section in spec.expected)


def _snake(section: str) -> str:
    return section.replace(" ", "_")


@dataclass
class ImprovementResult:
    """Typed improvement analysis for one artifact (ADR-003).

    Section names are stored in their normalized form (e.g. ``"success
    metrics"``); renderers do the casing. ``to_dict`` is the stable JSON
    contract (ADR-007): ``{type, missing_required, missing_recommended,
    guidance}`` with snake-cased section names.
    """

    type: str  # classified type, or "unknown"
    missing_required: list[str]
    missing_recommended: list[str]
    # Guidance questions for each missing section: {section -> questions}.
    guidance: dict[str, list[str]] = field(default_factory=dict)
    # Whether this type yields suggestions at all (spec present + full coverage).
    supported: bool = False
    # Reserved for future unknown-type handling (e.g. nearest match); not serialized.
    closest_type: str | None = None

    def to_dict(self) -> dict:
        return {
            "type": self.type,
            "missing_required": [_snake(s) for s in self.missing_required],
            "missing_recommended": [_snake(s) for s in self.missing_recommended],
            "guidance": {_snake(s): list(questions) for s, questions in self.guidance.items()},
        }


def improve_product(product: Product) -> ImprovementResult:
    """Analyze a parsed ``product`` and return its completion guidance."""
    artifact_type = classify(product).type
    spec = spec_for(artifact_type)
    if spec is None or not supports_improve(spec):
        # Unknown, or a known type whose schema lacks complete guidance: there is
        # nothing actionable to report.
        return ImprovementResult(
            type=artifact_type,
            missing_required=[],
            missing_recommended=[],
            supported=False,
        )

    missing_required, missing_recommended = missing_sections(product, spec)
    # Keep required sections ahead of recommended so renderers preserve priority.
    guidance = {
        section: list(spec.guidance[section])
        for section in missing_required + missing_recommended
        if spec.guidance.get(section)
    }
    return ImprovementResult(
        type=artifact_type,
        missing_required=missing_required,
        missing_recommended=missing_recommended,
        guidance=guidance,
        supported=True,
    )


def improve_text(text: str) -> ImprovementResult:
    return improve_product(parse(text))


def improve_file(path: str) -> ImprovementResult:
    return improve_product(parse_file(path))

File: rust/rac-engine/src/improve.rs
//! Artifact improvement (`decided.services.improve`): deterministic,
//! schema-driven guidance. Advisory and read-only — reports missing
//! required/recommended sections with schema-defined guidance questions.

use crate::classify::{classify, missing_sections};
use crate::parse::Artifact;
use crate::spec::{spec_for, ArtifactSpec};

/// `supports_improve(spec)` — every expected section defines guidance.
/// (All five current specs pass, so the unsupported branch is dead in
/// practice; ported for fidelity.)
pub fn supports_improve(spec: &ArtifactSpec) -> bool {
    spec.expected()
        .iter()
        .all(|section| spec.guidance.iter().any(|(k, _)| k == section))
}

/// Typed improvement analysis (`ImprovementResult`). Section names are
/// stored normalized; the renderers format them.
pub struct ImprovementResult {
    /// Classified artifact type, or `"unknown"`.
    pub artifact_type: String,
    pub missing_required: Vec<String>,
    pub missing_recommended: Vec<String>,
    /// Schema guidance for the missing sections: `{section -> questions}`,
    /// required-first then recommended, only sections that HAVE guidance.
    pub guidance: Vec<(String, Vec<String>)>,
    /// Whether `improve` produces suggestions for this type.
    pub supported: bool,
}

/// `improve_product(product)` — analyze and return improvement guidance.
pub fn improve_product(artifact: &Artifact) -> ImprovementResult {
    let artifact_type = classify(artifact).artifact_type;
    let spec = spec_for(&artifact_type);
    let Some(spec) = spec.filter(|s| supports_improve(s)) else {
        // Unknown, or a known type whose schema lacks complete guidance.
        return ImprovementResult {
            artifact_type,
            missing_required: Vec::new(),
            missing_recommended: Vec::new(),
            guidance: Vec::new(),
            supported: false,
        };
    };
    let (missing_required, missing_recommended) = missing_sections(artifact, spec);
    let mut guidance: Vec<(String, Vec<String>)> = Vec::new();
    for s in missing_required.iter().chain(missing_recommended.iter()) {
        // `if spec.guidance.get(s)` — only truthy (non-empty) guidance lists.
        if let Some((_, g)) = spec.guidance.iter().find(|(k, _)| k == s) {
            if !g.is_empty() {
                guidance.push((s.clone(), g.clone()));
            }
        }
    }
    ImprovementResult {
        artifact_type,
        missing_required,
        missing_recommended,
        guidance,
        supported: true,
    }
}
